src/export/ValueSetExporter.ts

```diff
@@ -130,6 +135,155 @@ export class ValueSetExporter {
+        )
+      ) {
+        errors.push('Composition contains filter operators.');
+      }
+      // including/excluding value sets are prohibited
+      if (
+        [...(valueSet.compose?.include ?? []), ...(valueSet.compose?.exclude ?? [])].some(
+          compose => {
+            return compose.valueSet?.length && !compose.concept?.length;
+          }
+        )
+      ) {
+        errors.push('Composition contains other value sets.');
+      }
+      // full usage of non-local code systems are prohibited
+      const referencedCodeSystems = fromPairs(
+        [...(valueSet.compose?.include ?? []), ...(valueSet.compose?.exclude ?? [])]
+          .filter(compose => {
+            return !compose.concept?.length && compose.system;
+          })
+          .map(compose => [
+            compose.system,
+            this.fisher.fishForFHIR(compose.system, Type.CodeSystem)
+          ])
+      );
+      // if we were able to fish up definitions for all of these systems, we're good
+      if (Object.values(referencedCodeSystems).includes(undefined)) {
+        errors.push('Composition contains code systems without available concept lists.');
+      }
+
+      // what we like best of all is when we just have a list of concepts. they'll even have a system specified already!
+      if (errors.length) {
+        logger.error(`Unable to expand ValueSet ${valueSet.name}: ${errors.join(' ')}`);
+      } else {
+        valueSet.expansion.contains = [];
+
+        valueSet.compose.include.forEach(compose => {
+          // add specific included codes
+          compose.concept?.forEach(concept => {
+            const containedItem: ValueSetExpansionContains = {
+              system: compose.system,
+              code: concept.code
+            };
+            if (compose.version) {
+              containedItem.version = compose.version;
+            }
+            if (concept.display) {
+              containedItem.display = concept.display;
+            }
+            if (concept.designation) {
+              containedItem.designation = cloneDeep(concept.designation);
+            }
+            // remove before adding, since we do not want it to be present multiple times
+            this.removeConcept(containedItem, valueSet.expansion.contains);
+            valueSet.expansion.contains.push(containedItem);
+          });
+          // add known included code systems
+          if (!compose.concept?.length && compose.system) {
+            const codeSystem = referencedCodeSystems[compose.system];
+            const newConcepts = this.getConcepts(codeSystem, codeSystem.concept);
+            // flatten so we can remove each existing concept before adding
+            const newConceptsFlat = flatMap(newConcepts, extractContained);
+            newConceptsFlat.forEach(concept =>
+              this.removeConcept(concept, valueSet.expansion.contains)
+            );
+            valueSet.expansion.contains.push(...newConcepts);
+          }
```

Comment:
This is, admittedly, a little bit weird, and merits some comments in order to explain the particulars of this ritual. I'll explain it as best I can here, so please let me know if it still seems ambiguous and/or incorrect.

When including an entire CodeSystem, the retrieved `CodeSystem.concept` array contains concepts that may, in turn, contain other concepts, forming a hierarchy. This hierarchy will be preserved when expanding the ValueSet.

Based on (http://hl7.org/fhir/R4/valueset.html#uniqueness), a concept should only appear once in the expansion. To avoid duplication, each concept present in the retrieved CodeSystem is removed from the expansion before `CodeSystem.concept` is added to the expansion. This is an implementation choice and not a specification requirement. It would be equally valid by the specification to instead implement this by instead performing a check on a concept before adding it, and remove it from `newConcepts` if it would result in a duplicate. I felt that "remove, then add" was simpler to implement than "check before adding," so that is the proposed implementation.

To explore the differences between the results of these two options for implementation, consider the [Allergy Intolerance](http://hl7.org/fhir/R4/codesystem-allergyintolerance-clinical.html) CodeSystem.

The FSH is formatted a little strangely due to markdown limitations, but is hopefully still understandable.

FSH | remove, then add | check before adding
--- | --- | ---
`* include codes from system AllergyIntolerance` | `[{code: 'active'}, {code: 'inactive', contains: [{code: 'resolved'}]}]` | `[{code: 'active'}, {code: 'inactive', contains: [{code: 'resolved'}]}]`
`* include AllergyIntolerance#resolved * include codes from system AllergyIntolerance` | `[{code: 'active'}, {code: 'inactive', contains: [{code: 'resolved'}]}]` | `[{code: 'resolved'}, {code: 'active'}, {code: 'inactive'}]`
`* include AllergyIntolerance#inactive * include codes from system AllergyIntolerance` | `[{code: 'active'}, {code: 'inactive', contains: [{code: 'resolved'}]}]` | `[{code: 'inactive'}, {code: 'active'}, {code: 'resolved'}]`

To explain line 218 a little more: `flatMap` maps the provided function, then flattens the resulting array. This flattening is only one layer deep, hence why the `extractContained` function also calls `flatMap` in order to end up with an array that contains all the concepts in a single flat list, with no more contained concepts in any of them. `extractContained` returns an array where the first element is the provided concept and the elements after that are the result of mapping `extractContained` on to each contained concept, then flattening each of those. So to walk through calling this with AllergyIntolerance, which starts off as `[{code: 'active'}, {code: 'inactive', contains: [{code: 'resolved'}]}]`

- call 1: `extractContained({code: 'active'})`: returns `[{code: 'active'}]`, since there is no `contains` property.
- call 2: `extractContained({code: 'inactive', contains: [{code: 'resolved'}]})`: returns `[{code: 'inactive'}, ...flatMap([{code: 'resolved'}], extractContained)]`
- - `extractContained({code: 'resolved'})`: returns `[{code: 'resolved'}]`, since there is no `contains` property.
- This is then flattened and spread, making the second call return `[{code: 'inactive'}, {code: 'resolved'}]`
- Back at the original `flatMap` call, the mapping is complete, and the array is `[[{code: 'active'}], [{code: 'inactive'}, {code: 'resolved'}]`. This is then flattened to produce `[{code: 'active'}, {code: 'inactive'}, {code: 'resolved'}]`.

Reply:
Thanks for the thorough explanation.  I think I understand what is going on then... Part of what confused me is that I thought you had said (at some point) that if you directly include a code (e.g., `resolved`) that it should bump the code to the top (the behavior of "check before adding").  But... I probably remembered that conversation wrong or it never happened at all.

For the second case above:
```
* include AllergyIntolerance#resolved
* include codes from system AllergyIntolerance
```
do you know what the expansion looks like when the IG Publisher expands it?  Does it look like the second column (w/ `resolve` still nested) or the third column (w/ `resolved` popped up to the top)?

Reply:
Alas! It seems that in this case, the IG publisher produces all three codes at the same level. That means that the "check before adding" implementation is the better match. I'll update the implementation accordingly.